Adds support to customize the sanitizer

dnsaoki2 opened the pull request:
Hi, this PR aims to solve #7710, adding the possibility to overwrite the ADD_TAGS and FORBID_TAGS options in sanitizer.

to overwrite these options, you need to configure the app-config.yaml file
``` yaml
techdocs:
   ....
   sanitizer:
      addTags:
          - tag1
          - tag2
      forbidTags:          
          - tag1
          - tag2
```

P.S.: I made a change that i'm not sure i should xD. The schema data filter when dealing with an array it checked the visibility of each item in the array. Is it expected having to declare the visibility of every position in the array? I think it`s a little weird having to declare visibility for every position in a string array, for example, but I might be missing some context here.

Rugvip commented on .changeset/eleven-wombats-glow.md:
```diff
@@ -0,0 +1,5 @@
+'@backstage/config-loader': minor
+---
+
+**BREAKING** Uses key visibility when the value isn't an array of objects in schema filters. So if the key of an array of string is visible, all elements of this array will also be visible.
```
Thank you for looking at this :grin: I actually think this could be a pretty good change, and that we could keep it backwards compatible as well by both looking at the array visibility path and the item visibility path as well.

One ask though, would you be alright with moving this `config-loader` change and opening up a separate PR for it? That way we can discuss and merge these things in isolation

emmaindal reviewed (commented):
Nice, from a techdocs perspective this looks good. I'm thinking that we could take the opportunity to add either a question and answer here https://backstage.io/docs/features/techdocs/faqs or a short how to guide here https://backstage.io/docs/features/techdocs/how-to-guides, do you want to write something? you can decide where you think it fits best! And then to @Rugvip's point, maybe separate the config loader change out to its own PR

dnsaoki2 commented:
Hi @emmaindal and @Rugvip , thanks for the feedback.

About the documentation, i will write something :) 

I will split the modification in config-loader in another PR, but one point is that the modification in sanitizer needs this modification to work ok?

dnsaoki2 commented:
@Rugvip I created a new PR with the change in config-loader and modified it to have backwards compatibility :smile: 

@emmaindal I added the documentation :smile:

emmaindal commented on docs/features/techdocs/how-to-guides.md:
````diff
@@ -421,3 +421,34 @@ folder (/docs) or replace the content in this file.
+```
+
+To overwrite the FORBID_TAGS configuration add this to your `app-config.yaml`.
+The default is ["sytle"]
````
```suggestion
The default is ["style"]
```

Signed-off-by: Emma Indal <emma.indahl@gmail.com>

emmaindal reviewed (commented):
Looks 💯, just one typo change from my side

ottosichert commented:
Hey @dnsaoki2, quick update from our side:

We are currently taking a deeper look into the security implications of allowing custom tags to be added to TechDocs pages on Backstage instances. Once we have a plan on how to move forward, we will come back to you in this PR or the linked issue!

emmaindal commented:
Hi @dnsaoki2 

We have discussed this a bit more and want to take a little bit of another approach. So thing that is requested here is to allow iframes. And we believe we can allow that but still in a secure way. So instead of allowing tags to be specified, we are thinking that we should allow iframe sources to be specified, and if thats specified we can allow iframes by default in the sanitizer. So something like: 

```
techdocs: 
    allowedIframeSources: ["https://drive.google.com"]
```

(open for suggestions on configuration terminology)

and when we sanitize the html, we can allow iframes if their source is allowed. 

The next step is to look into adding the sandbox attribute to the iframes and allow configuration for that too. We can do that in a separate PR though. 

Do you still want to continue look into this as a part of this PR?

dnsaoki2 commented:
Hi @emmaindal, thanks for the update.
yes i want to continue.

is it better to change this PR or create another one and close this one?

emmaindal commented:
> Hi @emmaindal, thanks for the update.
> 
> yes i want to continue.
> 
> 
> 
> is it better to change this PR or create another one and close this one?

Feel free to continue here if that works for you. Let us know if you have any follow up questions!

dnsaoki2 commented:
hi @emmaindal 
i updated the PR and made some changes regarding the approach.

I changed to be a list of allowed hosts, called allowedIframeHosts and put this setting inside the 'sanitizer' key in techdocs.

something like:
```yaml
techdocs:
   sanitizer:
      allowedIframeHosts:
         - drive.google.com
```
what do you think of this approach?

Rugvip commented on plugins/techdocs/src/reader/transformers/sanitizeDOM.ts:
```diff
@@ -34,14 +34,51 @@ export const safeLinksHook = (node: Element) => {
+        'beforeSanitizeElements',
+        filterIframeHook(allowedIframeHosts),
+      );
+      addTags.push('iframe');
```
Is there any default sanitation that's applied to the iframe elements? It's very important that we filter out vulnerable attributes like `onload`. If it isn't already done it would be preferable if it's done with a 3rd party library, explicit allow list, or based on solid and referenced security recommendations.

Rugvip replied:
Alright, good. Could you add some tests for those cases as well? Another important one is `srcdoc`, so we wanna make sure that a valid `src` does not allow an element with a `srcdoc` through for example

dnsaoki2 replied:
@Rugvip test added :)

emmaindal commented on docs/features/techdocs/how-to-guides.md:
```diff
@@ -421,3 +421,24 @@ folder (/docs) or replace the content in this file.
 
 Done! You now have support for TechDocs in your own software template!
+
+## how to enable iframes in TechDocs
```
thank you for writing this how to guide too!

emmaindal commented on plugins/techdocs/src/reader/transformers/sanitizeDOM.test.ts:
```diff
@@ -111,6 +112,108 @@ describe('sanitizeDOM', () => {
+      postTransformers: [],
+    });
+    expect(shadowDom.querySelectorAll('link').length).toEqual(1);
+    expect(shadowDom.querySelectorAll('iframe').length).toEqual(1);
```
can we add an assertion here to make sure the src is equal to `https://example.com?test=1` as well?

dnsaoki2 replied:
@emmaindal done :)

emmaindal replied:
⭐

emmaindal reviewed (commented):
Thanks a lot for looking into this @dnsaoki2, just one small comment then I think we can merge this if @Rugvip does not have any additional input?

emmaindal reviewed (approved):
Thanks, approval from me! @Rugvip good to go?

emmaindal commented:
@dnsaoki2 please rebase this one master and the e2e test should pass!

emmaindal commented on plugins/techdocs/src/reader/transformers/sanitizeDOM.ts:
```diff
@@ -34,14 +34,51 @@ export const safeLinksHook = (node: Element) => {
+
+    try {
+      const srcUrl = new URL(src);
+      const isMatch = allowedIframeHosts.some(host => srcUrl.host === host);
```
one thing that just came to mind, I wonder if we should allow hosts specified with wildcards in the domain. For example something like `.*.google.com` and then we can do something like 

```ts
allowedIframeHosts.some(host => srcUrl.host.match(host));
```

What do you think @Rugvip @dnsaoki2 ?

dnsaoki2 replied:
@emmaindal I don't know if there would be any security issues, a point of attention would be to pass for example ".*" as allowed host, this will basically allow any iframe

emmaindal replied:
that is a very good point, thanks for brining that up! Let's not support that for now and keep it to the way it is!

Rugvip replied:
I'd say let's not add it unless it turns out we really need it tbh, bit too sensitive area. I'd also steer clear of regex for this kind of matching too, as for example `xyzabcgoogle.com` matches the example pattern

emmaindal commented:
Thanks again @dnsaoki2 for you awesome contribution here! ⭐

dnsaoki2 commented:
Thanks @emmaindal  and @Rugvip  for your attention and patience :smile:
